staging/src/k8s.io/apiserver/pkg/cel/template/template.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package template
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common"
+	"github.com/google/cel-go/common/types"
+
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/cel/environment"
+)
+
+const (
+	// ObjectName is the variable name for the input data in a CEL expression.
+	ObjectName = "object"
+)
+
+var (
+	placeholderRe = regexp.MustCompile(`{{.*?}}`)
+	stringType    = reflect.TypeOf("")
+)
+
+// NewTemplate parses the template string and all CEL expressions embedded in
+// it. CEL expressions are enclosed with {{<expression>}}. That placeholder
+// gets replaced by the value of that expression. The value must be a string
+// or support conversion to a string.
+//
+// If there are parse errors, then the returned error contains one error for
+// each expression that failed to parse. All errors from the underlying CEL
+// libraries are wrapped.
+//
+// Parsing is more permissive when using [environment.StoredExpression] as
+// environment type. Use [environment.NewExpression] when dealing with
+// templates before persisting them to storage.
+func NewTemplate(template string, options ...Option) (*Template, error) {
+	// TODO: move to init() to save repeated costs. Errors become panics?!
+	ver := environment.DefaultCompatibilityVersion()
+	envSet := environment.MustBaseEnvSet(ver)
+
+	objectOpt := cel.Variable(ObjectName, types.MapType)
+	versionedObjectOpt := environment.VersionedOptions{
+		IntroducedVersion: version.MajorMinor(1, 30),
+		EnvOptions:        []cel.EnvOption{objectOpt},
+	}
+	envSet, err := envSet.Extend(versionedObjectOpt)
+	if err != nil {
+		return nil, fmt.Errorf("extending CEL environment with object variable: %w", err)
+	}
+
+	c := config{
+		envType: environment.NewExpressions,
+	}
+	for _, opt := range options {
+		if err := opt(&c); err != nil {
+			return nil, err
+		}
+	}
+	env, err := envSet.Env(c.envType)
```

Comment:
you have to add the variables ("object") into this envirnoment

Reply:
Isn't that what this code does (`ObjectName` = "object")?

```
	objectOpt := cel.Variable(ObjectName, types.MapType)
	versionedObjectOpt := environment.VersionedOptions{
		IntroducedVersion: version.MajorMinor(1, 30),
		EnvOptions:        []cel.EnvOption{objectOpt},
	}
	envSet, err := envSet.Extend(versionedObjectOpt)
```

I cannot add the actual variable value yet. That's going to get passed in when evaluating the program.

Reply:
yes, sorry, shouldn't review during a meeting...I missed that

I will give it a run later today, if I have time. I do see you are also passing in the object later during eval, I believe.

Reply:
Where I am uncertain is whether `types.MapType` is correct for the `types.NewStringInterfaceMap(t.env.CELTypeAdapter(), object)` value. But I am not even getting that far, it's already the expression parsing which fails.

Reply:
I can see in a debugger that `env.variables` is indeed empty. It seems to get dropped somewhere. I'll double-check...

Reply:
It's set for `environment.StoredExpression`, but not for `environment.NewExpression`. Something with `environment.VersionedOptions` isn't quite right.

Reply:
I think I understand the underlying rationale: to use the new variable, support for it would have had to be present already in 1.29 for safe rollbacks.

This is not applicable here (much more than just CEL evaluation will break when downgrading), so perhaps fudging it by claiming that 1.29 already had "object" is okay.